fix(chat): Fix the run_command tool on Windows (#2179)

alzwded opened the pull request:
## Description

<!-- If this is a feature request, please describe how it will benefit other CodeCompanion users -->
Parameter quoting rules on Windows are complicated, because Windows programs only receive one big string with all arguments which they or msvcrt split back into argv. libuv used by NeoVim tries to be helpful, but ends up sending literal quotes to cmd.exe (programs get `"C:\file.txt"` with quotes instead of `C:\file.txt`) or not literal enough to powershell (literal empty arguments `""` get lost along the way)

For those who don't doze on their wind: https://web.archive.org/web/20190109172835/https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/

Creating a temporary .bat file and executing it _almost_ works, but I have had the LLM propose running `for ... %i ...` commands which require doubling up the percentage sign. Even with prompting, it would still occasionally throw a `FOR` curveball.

What does work, is typing the command INTO cmd.exe, and letting cmd.exe figure things out. So we send table.concat(cmd, " ") on stdin, and run cmd.exe. But... we need a couple of things set up so as to not confuse the the LLM:
* pass /K to inhibit the Microsoft Logo and copyright; it will exit when stdin is exhausted
* pass /Q to inhibit showing the command we're executing, since that's not part of the output
* set %PROMPT% to a space, so that it doesn't see any c:\> prompt; %PROMPT% == "" means "use the default", hence one space
* tac a "\n" at the end of the command, since cmd doesn't treat EOF as an implicit CR, and it would just print out `More?` like a silly goose.

This seems to have resolved all the issues I was encountering.

I've added a windows-only test which invokes the `find` command (no relation to POSIX `find`), which is one of the more troublesome commands. 

## Related Issue(s)

<!--
  If this PR fixes any issues, please link to the issue here.
  - Fixes #<issue_number>
-->
- Fixes #2179 

## Screenshots

<!-- Add screenshots of the changes if applicable, to help visualize the change. -->
N/A

## Checklist

- [x] I've read the [contributing](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md) guidelines and have adhered to them in this PR
- [x] I've added [test](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md#testing) coverage for this fix/feature
- [x] I've run `make all` to ensure docs are generated, tests pass and my formatting is applied
- [ ] _(optional)_ I've updated `CodeCompanion.has` in the [init.lua](https://github.com/olimorris/codecompanion.nvim/blob/main/lua/codecompanion/init.lua#L239) file for my new feature
- [ ] _(optional)_ I've updated the README and/or relevant docs pages

alzwded commented:
Mmm... actually, tests seem to be failing, either because they now get an extra `^M`, or, even more strangely, in WSL they get a full PROMPT in the output. Argh...

olimorris commented on lua/codecompanion/strategies/chat/tools/orchestrator.lua:
```diff
@@ -69,7 +69,46 @@ local function cmd_to_func_tool(tool)
+          end
+
+          if vim.fn.has("win32") then
+            -- Parameter quoting rules on Windows are complicated, because
```
Let's remove this comment block and just reference this PR number

olimorris commented:
Firstly, thanks so much for adding test coverage. Really, really appreciate that.

Few minor points but this looks great.

alzwded commented:
> Firstly, thanks so much for adding test coverage. Really, really appreciate that.
> 
> Few minor points but this looks great.

No probs. I'll sort out the comments. I'm wondering if that test needs to be in a different suite?

I actually see test failures in very strange ways (I see CI also failed here) on test_cmd in particular. Like... when running in WSL, it somehow ends up calling cmd.exe, even though when I run nvim in there it `has('win32')` reports false. This is going to take a while....

alzwded commented on lua/codecompanion/strategies/chat/tools/orchestrator.lua:
```diff
@@ -37,7 +37,7 @@ local function cmd_to_func_tool(tool)
         return function(tools, _, _, cb)
           cb = vim.schedule_wrap(cb)
-          vim.system(tool_utils.build_shell_command(cmd), {}, function(out)
+          local out_cb = function(out)
```
Yes. Referencing `vim.system`, I've renamed `out_cb` to `system_on_exit_cb`, `out` to `process_result_obj` (the docs' example just call it obj), and `pattern` to `split_pattern`.

olimorris commented on tests/strategies/chat/tools/runtime/test_cmd.lua:
I'd really like to not increase these waits. They take long enough to run on my machine as it is. If this is a Windows fix, then let's detect it and adjust the waits up front.

alzwded replied:
Tell you what -- I've left out that commit. I'm thinking of doing a different PR to make tests run on native Windows, as a whole. I'll keep what you said in mind and if timeouts need to be adjusted for Windows only, it will be for windows only. "Fixing" tests on native Windows isn't strictly related to my Bug report...

olimorris commented:
Awesome. Just a minor fix regarding the wait times then I'll be happy to merge this. Thanks.

olimorris commented:
@alzwded - I've split out the shell execution into its own func. Tests pass my end. Can you check they're good for you?

alzwded commented:
> > @alzwded - I've split out the shell execution into its own func. Tests pass my end. Can you check they're good for you?
> 
> @olimorris I've checked again, and in the callback you pass to execute shell command, it's rather important to use `"\r?\n"` on Windows, since lines tend to be terminated by `\r\n` (unless some file or some program refuses to throw in ASCII 13 in there; which is why I set it to _LF maybe preceded by CR_).
> 
> I've added a commit to add the difference in EOL pattern on Windows.

For context, without my `"\r?\n"` commit, this happens:

<describe>
<summary>Failing tests</summary>

```
FAIL in tests\strategies\chat\tools\catalog\test_cmd_runner.lua | cmd_runner tool:
  Failed expectation for screenshot equality to reference at "tests/screenshots/tests-strategies-chat-tools-catalog-test_cmd_runner.lua---cmd_runner-tool".
  Different `text` cell at line 9 column 1. Reference: "h". Observed: "^".
  Reference:
  --|---------|---------|---------|---------|---------|---------|---------|---------|
  01|## foo                              â",
  02|                                    â",~
  03|                                    â",~
  04|                                    â",~
  05|## assistant                        â",~
  06|                                    â",~
  07|`echo hello world`                  â",~
  08|```                                 â",~
  09|hello world                         â",~
  10|```                                 â",~
  11|                                    â",~
  12|## foo                              â",~
  13|                                    â",~
  14|                                    â",~
  15|~                                   â",~
  16|~                                   â",~
  17|~                                   â",~
  18|~                                   â",~
  19|~                                   â",~
  20|~                                   â",~
  21|~                                   â",~
  22|~                                   â",~
  23|~                                   â",~
  24|~                                   â",~

  --|---------|---------|---------|---------|---------|---------|---------|---------|
  01|00000011111111111111111111111111111121111111111111111111111111111111111111111111
  02|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  03|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  04|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  05|00000000000011111111111111111111111123333333333333333333333333333333333333333333
  06|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  07|44444444444444444411111111111111111123333333333333333333333333333333333333333333
  08|44411111111111111111111111111111111123333333333333333333333333333333333333333333
  09|44444444444111111111111111111111111123333333333333333333333333333333333333333333
  10|44411111111111111111111111111111111123333333333333333333333333333333333333333333
  11|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  12|00000011111111111111111111111111111123333333333333333333333333333333333333333333
  13|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  14|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  15|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  16|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  17|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  18|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  19|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  20|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  21|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  22|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  23|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  24|33333333333333333333333333333333333323333333333333333333333333333333333333333333

  Observed:
  --|---------|---------|---------|---------|---------|---------|---------|---------|
  01|## foo                              â",
  02|                                    â",~
  03|                                    â",~
  04|                                    â",~
  05|## assistant                        â",~
  06|                                    â",~
  07|`echo hello world`                  â",~
  08|```                                 â",~
  09|^M                                  â",~
  10|hello world^M                       â",~
  11|^M                                  â",~
  12|^M                                  â",~
  13|```                                 â",~
  14|                                    â",~
  15|## foo                              â",~
  16|                                    â",~
  17|                                    â",~
  18|~                                   â",~
  19|~                                   â",~
  20|~                                   â",~
  21|~                                   â",~
  22|~                                   â",~
  23|~                                   â",~
  24|~                                   â",~

  --|---------|---------|---------|---------|---------|---------|---------|---------|
  01|00000011111111111111111111111111111121111111111111111111111111111111111111111111
  02|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  03|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  04|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  05|00000000000011111111111111111111111123333333333333333333333333333333333333333333
  06|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  07|44444444444444444411111111111111111123333333333333333333333333333333333333333333
  08|44411111111111111111111111111111111123333333333333333333333333333333333333333333
  09|55111111111111111111111111111111111123333333333333333333333333333333333333333333
  10|44444444444551111111111111111111111123333333333333333333333333333333333333333333
  11|55111111111111111111111111111111111123333333333333333333333333333333333333333333
  12|55111111111111111111111111111111111123333333333333333333333333333333333333333333
  13|44411111111111111111111111111111111123333333333333333333333333333333333333333333
  14|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  15|00000011111111111111111111111111111123333333333333333333333333333333333333333333
  16|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  17|11111111111111111111111111111111111123333333333333333333333333333333333333333333
  18|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  19|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  20|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  21|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  22|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  23|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  24|33333333333333333333333333333333333323333333333333333333333333333333333333333333
  Traceback:
    tests\strategies\chat\tools\catalog\test_cmd_runner.lua:40

FAIL in tests\strategies\chat\tools\runtime\test_cmd.lua | Tools | cmds | handlers and outputs are called:
  Failed expectation for equality.
  Left:  "Hello World"
  Right: "\r"
  Traceback:
    tests\strategies\chat\tools\runtime\test_cmd.lua:56

FAIL in tests\strategies\chat\tools\runtime\test_cmd.lua | Tools | cmds | can run multiple commands:
  Failed expectation for equality.
  Left:  { "Hello World" }
  Right: { "\r", "Hello World\r", "\r", "\r" }
  Traceback:
    tests\strategies\chat\tools\runtime\test_cmd.lua:107


```

</describe>

With the `\r?\n` thing, those tests pass, since the `\r`s get stripped together with the `\n`.

I didn't notice the LLM getting confused by stray CRs, but it's probably better to keep things consistent across platforms.
